ros/src/tl_detector/light_classification/tl_classifier.py

```diff
@@ -1,78 +1,160 @@
+
+    INSTANCE = None
+    KNOWN_TRAFFIC_LIGHT_CLASSIFIERS = {}  # it is not empty; it is filled by TLClassifier.register_subclass decorator
+
+    @classmethod
+    def register_subclass(cls, cls_id):
+        """
+        Decorator for TLClassifier subclasses.
+        Adds decorated class to the cls.KNOWN_TRAFFIC_LIGHT_CLASSIFIERS dictionary.
+        :param cls_id: string identifier of the classifier
+        :return: function object
+        """
+        def reg_subclass(cls_type):
+            cls.KNOWN_TRAFFIC_LIGHT_CLASSIFIERS[cls_id] = cls_type
+            return cls_type
+        return reg_subclass
+
+    @classmethod
+    def get_instance_of(cls, classifier_name):
+        """
+        It is a factory method for the `tl_classifier` module. It returns an instance of the classifier
+        based on the input argument provided.
+        :param classifier_name: name of the classifier
+        :type classifier_name: str
+        :return: instance of the classifier corresponding to the classifier string identifier
+        :rtype: TLClassifier
+        """
+        if cls.INSTANCE is not None \
+                and type(cls.INSTANCE) != cls.KNOWN_TRAFFIC_LIGHT_CLASSIFIERS[classifier_name]:
+            raise ValueError("cannot instantiate an instance of " + classifier_name
+                             + " classifier since an instance of another type (" + type(cls.INSTANCE).__name__ +
+                             ") has already been instantiated")
+
+        if cls.INSTANCE is not None:
+            return cls.INSTANCE
+
+        classifier_type = cls.KNOWN_TRAFFIC_LIGHT_CLASSIFIERS.get(classifier_name, None)
+        if classifier_type is None:
+            raise ValueError("classifier_name parameter has unknown value: " + classifier_name
+                             + "; the value should be in " + str(cls.KNOWN_TRAFFIC_LIGHT_CLASSIFIERS.keys()))
+        cls.INSTANCE = classifier_type()
+
+        return cls.INSTANCE
+
+    @abstractmethod
+    def _classify(self, image):
+        """
+        Determines the color of the traffic light in the image.
+        This method should be implemented by a particular type of the traffic light classifier.
+
+        :param image: image containing the traffic light
+        :type image: np.ndarray
+        :returns: ID of traffic light color (specified in styx_msgs/TrafficLight)
+        :rtype: int
+        """
+        raise NotImplementedError()
+
+    def classify(self, image):
+        """
+        Determines the color of the traffic light in the image.
+        Prints FPS statistic approximately each second.
+
+        :param image: image containing the traffic light
+        :type image: np.ndarray
+        :returns: ID of traffic light color (specified in styx_msgs/TrafficLight)
+        :rtype: int
+        """
+        # calculate FPS based on the number of images processed per second;
+        # ensure that self._counter value does not go over integer limits
+        with self._lock:
```

Comment:
why do u need to acquire a lock before you can run the code?  Is this blocking wait?

Reply:
usually you will need locks if there are multiple threads trying to access a shared resource, i haven't  used threads in python for a long time, so ...

Reply:
TLClassifier.classify is called from the callback of some Subscriber in TLDetector. Callback may be called in parallel by ROS subscribers. `+=` is not an atomic operation in Python, thus without lock self._counter may miss some of the increments. Yes, it is a blocking wait. Critical section has 1 condition check, one increment and 2 assignments, so, threads do not spend much time in the critical section.